### PR TITLE
feat: Enable basic sandboxed tool run functionality

### PR DESCRIPTION
**Pull request by carenthomas**
**Please describe the purpose of this pull request.**
Basic functionality to support running user-specified tools in sandbox. Submitting v1 for feedback on approach! 

**How to test**
How can we test your PR during review? What commands should we run? What outcomes should we expect?
```
poetry run pytest -s -vv tests/test_tools.py::test_run_basic_tool_in_sandbox
poetry run pytest -s -vv tests/test_tools.py::test_run_tool_with_params_in_sandbox
```

**Have you tested this PR?**
Have you tested the latest commit on the PR? If so please provide outputs from your tests.

**Related issues or PRs**
Please link any related GitHub [issues](https://github.com/cpacker/Letta/issues) or [PRs](https://github.com/cpacker/Letta/pulls).

**Is your PR over 500 lines of code?**
If so, please break up your PR into multiple smaller PRs so that we can review them quickly, or provide justification for its length.

**Additional context**
Add any other context or screenshots about the PR here.

### CODE REVIEW

File: letta/agent.py [hunk 1]
```diff
@@ -651,6 +652,48 @@ def _handle_ai_response(
...
+                    code += ")"
+                    
+                    # 6. Execute code and extract result or throw
+                    execution = sbx.run_code(code)
+                    if execution.error is not None:
+                        raise execution.error
+                    elif len(execution.results) == 0:
+                        function_response = ""
+                    else:
+                        function_response = execution.results[0].text
```
Comment (sarahwooders) [inline letta/agent.py:690]: is there any issue with typing here - e.g. if the function returns a list? does this assume the function is always returning a string?

Comment (carenthomas) [reply letta/agent.py:690]: e2b will return lists as a string and store it in the text field! i.e.
```
python3
>>> from e2b_code_interpreter import Sandbox
>>> sbx = Sandbox(api_key=<API_KEY>)
>>> e = sbx.run_code("""
... def func(a):
...     return [1] * a
... 
... func(5)
... """)
>>> e
Execution(Results: [Result([1, 1, 1, 1, 1])], Logs: Logs(stdout: [], stderr: []), Error: None)
>>> e.results[0].text
'[1, 1, 1, 1, 1]'
```
It seems like the primitive types will show up here, but there are more complex types that have their own [field](https://github.com/e2b-dev/code-interpreter/blob/main/python/e2b_code_interpreter/models.py#L94-L122). Should be straightforward to extend as needed in the future.

**Review by carenthomas — COMMENTED** [review]
Fixed my trailing todos and added some helpers! 

I got the api key working as well locally with a really hacky set up so will clean that up next (specifically, calling client.create_tool using the exec(Tool.get_composio_tool(...).source_code) way makes the lookup for inspect.getsource fail) or just hope that @mattzh72's refactor with the add_tool magically makes things work!

Hope that @mattzh72's refactor consolidating the client.add_tool magically makes everything work!

**Comment by mattzh72** [comment]
Duplicated by #2040

**Pull request closed without merging**